core/src/main/java/org/apache/druid/math/expr/Expr.java

```diff
@@ -118,11 +182,41 @@ public ExprEval eval(ObjectBinding bindings)
   }
 }
 
+class LongArrayExpr extends ConstantArrayExpr
+{
+  private final Long[] value;
```

Comment:
These numbers are object types currently because that was the path of least resistance to getting this thing functioning and haven't spent much time optimizing yet, this is a good question though and worth discussing I think.

I do currently allow nulls as elements because right now all array inputs are either coming from multi-value string dimensions which must be cast or `string_to_array` which also is currently spitting out a string array and also needs cast (side note, i should probably allow `string_to_array` to specify an array type).

That said, even with nulls it would probably be better to use primitive arrays for the numeric types and keep a bitset of which elements are null similar to what we are doing for numerical columns. Alternatively we could distinguish between nullable array types and non-null array types. I have a slight preference for the former approach so we aren't drowning in types.

I'll play around and see what I can do.

Reply:
hmmm, allowing nulls in string array sounds ok but allowing them in long and double array types is somewhat unintuitive . I probably don't understand all the nuances , so feel free to do what makes most sense, we can revise things later if needed.

Reply:
If `druid.generic.useDefaultValueForNull` is false, it's possible to have null longs and doubles, so I think it makes sense to allow that for arrays as well

Reply:
I have been thinking about this a bit more, and I think I would rather keep this as it currently is with number objects instead of primitives and relaxed null handling in favor of revisiting optimizing expressions as a whole as a future effort. I can definitely be convinced otherwise though, so if anyone feels strongly feel free to make a case.

My primary reasoning for this is that (out of the box at least) all arrays we will encounter currently will be coming from multi-value string dimensions, so they are going to start their life and end their life as string objects if they stay as arrays with the current implementation. Further, expression evaluation is not currently fully optimized to use and take advantage of primitive types end to end during evaluation, which might have motivated me a bit more towards using primitives out of the gate.

I think to achieve optimal expression evaluating selectors we will first need to enrich the type information available throughout the query pipeline so that selector typing information that creates the input bindings is always complete (at least for queries). I referenced this as potential follow up work to this PR, since it would also be beneficial to this PR because it would reduce or eliminate the reliance on `RowBasedExpressionColumnValueSelector`. With complete typing information on input selectors in place, we can examine the complete `Expr` and infer output types for the entire expression tree, and use that to optimize how the expression is evaluated .. _somehow_. Still connecting some dots on what the best approach would be, I think there is a lot to consider. There is definitely a tedious manual way, but it may be that the best solution might involve some sort of code generation due to the shear number of combinations of things that are probably possible.